### PR TITLE
[VS Incentives] test/refactor: incentive epoch hook test for group; bug fixes

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #6526

## What is the purpose of the change

This PR tests x/incentives `AfterEpochEnd` hook with groups configured. This is a comprehensive test covering perpetual and non-perpetual group distributing across several epochs.

After every epoch, we assert that the underlying group gauges receive the expected amounts. Additionally, we assert that the scheduling policy is correct and pruning for non-perpetual functions as expected

## Bug fixes

There were 2 bugs caught:
1. The group in `AllocateAcrossGauges` was not refetched. As a result, the volume wasn't synched correctly. The tests didn't catch this because the input gauge always had non-zero gauge weights. Test to be expanded in a follow-up
2. Another bug was due to getting division by zero in `AllocateAcrossGauges` when volume delta ended up being zero. Fix is added. To be tested in a follow-up

## Drive By

- Added some `osmoutils` helpers that might be reused in the future
- Cleaned up and reused some test helpers

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes?
  - [ ] Changelog entry added to `Unreleased` section of `CHANGELOG.md`?

Where is the change documented? 
  - [ ] Specification (`x/{module}/README.md`)
  - [ ] Osmosis documentation site
  - [ ] Code comments?
  - [ ] N/A

### CODE REVIEW

**Comment by p0mvn** [comment]
@pysel I don't think the go.mod update tool is working correctly. See this commit: [3942961](https://github.com/osmosis-labs/osmosis/pull/6548/commits/3942961b35b77aa713b003173dae3d739eb22bb6)

Note that I made changes to `osmoutils` in this PR. As a result, I would expect `the go.mod tool to bump up osmoutils dependency here

**Comment by pysel** [comment]
Weird, I will take a look

**Comment by pysel** [comment]
I think I got it, I will open an issue and attach it to this comment in a bit

File: x/incentives/keeper/distribute_test.go [hunk 1]
```diff
@@ -1560,6 +1560,7 @@ func (s *KeeperTestSuite) TestSyncGroupWeights() {
 // - checking that groups are skipped in acceptable cases without failing other group distributions
 // - checking that the group gauge is correctly updated - filled epochs increased and distributed coins increased
 // - checking that the internal gauges are correctly updated - coins increased
+// TODO: add test where group does not have weights synched (empty)
```
Comment (p0mvn) [inline x/incentives/keeper/distribute_test.go:1563]: Note: TODO

File: x/incentives/keeper/distribute.go [hunk 2]
```diff
@@ -41,6 +42,14 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context, activeGroups []types.Group
 			continue
 		}
 
+		// Refetch group
+		// TODO: consider mutating receiver of syncGroupWeights instead of refetching.
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:46]: NOTE: TODO

File: x/incentives/keeper/distribute.go [hunk 3]
```diff
@@ -41,6 +42,14 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context, activeGroups []types.Group
...
+		// TODO: cover AllocateAcrossGauges with unit tests. to make sure that the bug is caught.
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:47]: Note: TODO

File: osmoutils/coin_helper_test.go [hunk 4]
```diff
@@ -252,3 +253,41 @@ func TestConvertCoinsToDecCoins(t *testing.T) {
 		})
 	}
 }
+
+func TestMergeCoinMaps(t *testing.T) {
```
Comment (p0mvn) [inline osmoutils/coin_helper_test.go:257]: Note: sponsored by chat gpt, thus, non-standard syntax

File: x/incentives/keeper/distribute.go [hunk 5]
```diff
@@ -82,6 +93,7 @@ func (k Keeper) AllocateAcrossGauges(ctx sdk.Context, activeGroups []types.Group
 		// Iterate over underlying gauge records in the group.
 		for gaugeIndex, distrRecord := range group.InternalGaugeInfo.GaugeRecords {
 			// Between 0 and 1. to determine the pro-rata share of the total amount to distribute
+			// TODO: handle division by zero gracefully and update test
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:96]: handle division by zero gracefully in AllocateAcrossGauges and update test

File: x/incentives/keeper/hooks_test.go [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+// Test_Distribute_Group_CreateGroupsBetween
+
+// TODO: create test
+// Test_Distribute_Group_SwapAndDistribute
+
+// TODO: create test
+// Test_Distribute_Group_ToggleVolumeAcrossEpochs
+
+// TODO: create test
+// Test_Distribute_Group_ToggleVolumeAcrossEpochs
```
Comment (p0mvn) [inline x/incentives/keeper/hooks_test.go:200]: Note tracked: more CreateGroup and Distribute test cases

File: x/incentives/keeper/distribute.go [hunk 7]
```diff
@@ -464,6 +477,14 @@ func (k Keeper) syncVolumeSplitGroup(ctx sdk.Context, group types.Group) error {
 			return types.CumulativeVolumeDecreasedError{PoolId: poolId, PreviousVolume: gaugeRecord.CumulativeWeight, NewVolume: cumulativePoolVolume}
 		}
 
+		// TODO: update this error and cover by tests
+		// If this is not present, getting a division by zero panic
+		// See:
+		// https://github.com/osmosis-labs/osmosis/issues/6558
+		if volumeDelta.IsZero() {
+			return errors.New("volume delta is zero")
+		}
```
Comment (p0mvn) [inline x/incentives/keeper/distribute.go:486]: Note: tracked https://github.com/osmosis-labs/osmosis/issues/6558

File: x/incentives/keeper/hooks_test.go [hunk 8]
```diff
@@ -0,0 +1,287 @@
...
+		s.Require().Equal(poolIDToExpectedDistributionMapOne[poolID].String(), gauge.Coins.String())
+	}
+}
+
+func (*KeeperTestSuite) computeExpectedDistributonAmountsFromVolume(poolIDToVolumeMap map[uint64]math.Int, totalVolume math.LegacyDec) map[uint64]sdk.Coins {
+	poolIDToExpectedDistributionMapOne := map[uint64]sdk.Coins{}
+	for poolID, volume := range poolIDToVolumeMap {
+		currentDistribution := coins.MulDec(defaultCoins, volume.ToLegacyDec().Quo(totalVolume))
+
+		fmt.Printf("poolId %d, currentDistribution %s\n", poolID, currentDistribution)
```
Comment (p0mvn) [inline x/incentives/keeper/hooks_test.go:227]: Note; Propose leaving this here as it is useful for debugging

File: x/incentives/keeper/distribute_test.go [hunk 9]
```diff
@@ -1583,9 +1583,11 @@ func (s *KeeperTestSuite) TestAllocateAcrossGauges() {
 		defaultGroup      = deepCopyGroup(defaultGroup)
 		singleRecordGroup = deepCopyGroup(singleRecordGroup)
 
+		two = sdk.NewInt(2)
```
Comment (czarcas7ic) [inline x/incentives/keeper/distribute_test.go:1586]: why multiply everything by two here?

File: x/incentives/keeper/hooks_test.go [hunk 10]
```diff
@@ -0,0 +1,292 @@
...
+	err = s.App.IncentivesKeeper.AddToGaugeRewardsInternal(s.Ctx, defaultCoins, perpetualGroupGaugeID)
+	s.Require().NoError(err)
+
+	// System under test
+	err = s.App.IncentivesKeeper.AfterEpochEnd(s.Ctx, distrEpochIdentifier, 3)
+	s.Require().NoError(err)
+
+	// Validate distribution
+
+	// Note that the perpetual gauge was not refunded. As a result, it is not distributing anymore.
```
Comment (czarcas7ic) [inline x/incentives/keeper/hooks_test.go:155]: I think this comment is not valid here

File: x/incentives/keeper/hooks_test.go [hunk 11]
```diff
@@ -0,0 +1,292 @@
...
+		s.Require().Equal(poolIDToExpectedDistributionMapOne[poolID].String(), gauge.Coins.String())
+	}
+}
+
+// computes the expected distribution values for each pool in the map based on the volume each one has and the total volume.
+// The expected distribution is calculated pro-rata based on the volume of each pool.
+func (*KeeperTestSuite) computeExpectedDistributonAmountsFromVolume(poolIDToVolumeMap map[uint64]math.Int, totalVolume math.LegacyDec) map[uint64]sdk.Coins {
+	poolIDToExpectedDistributionMapOne := map[uint64]sdk.Coins{}
+	for poolID, volume := range poolIDToVolumeMap {
+		currentDistribution := coins.MulDec(defaultCoins, volume.ToLegacyDec().Quo(totalVolume))
```
Comment (czarcas7ic) [inline x/incentives/keeper/hooks_test.go:229]: do you think it should be called something different than coins, maybe coinsutil or something? I embarrassingly was looking around for where the coins object was and finally realized it was just an import lol.

**Review by czarcas7ic — APPROVED** [review]
Nice job on the hooks_test

File: x/incentives/keeper/hooks_test.go (hunk 10 above)
Comment (p0mvn) [reply x/incentives/keeper/hooks_test.go]: fixed

File: x/incentives/keeper/hooks_test.go (hunk 11 above)
Comment (p0mvn) [reply x/incentives/keeper/hooks_test.go:229]: Fair point, I think it's outside of the scope of this PR. Will make an issue

Comment (p0mvn) [reply x/incentives/keeper/hooks_test.go:229]: https://github.com/osmosis-labs/osmosis/issues/6583

File: x/incentives/keeper/distribute_test.go (hunk 9 above)
Comment (p0mvn) [reply x/incentives/keeper/distribute_test.go:1586]: Added comment:
```
// Double the volume configuration in poolmanager because we want the current volume to be
// updated relative to the existing values in gauge record state.
// The current volume is computed = poolmanager cumulative volume - gauge record cumulative volume.
```
